### PR TITLE
feat: add sar teacher forcing

### LINKED ISSUE #71: [Models] Add SAR teacher forçing during training (decoder)
Teacher forçing (feeding the LSTM decoder of SAR with ground-truth characters during training) is not implemented, and it should really improve performances

### PR DESCRIPTION
**Pull request by charlesmindee**
Successfully merging this PR closes #71 , any feedback is welcome!

### CODE REVIEW

File: doctr/models/recognition/sar.py [hunk 1]
```diff
@@ -141,7 +141,13 @@ def call(
             # shape (N, rnn_units + 1) -> (N, vocab_size + 1)
             logits = self.output_dense(logits, **kwargs)
             # update symbol with predicted logits for t+1 step
-            symbol = tf.argmax(logits, axis=-1)
+            if labels:
```
Comment (fg-mindee) [inline doctr/models/recognition/sar.py:144]: Let's make this test more robust:
```
if labels is not None:
```
otherwise passing a boolean = True would produce errors

File: doctr/models/recognition/sar.py [hunk 2]
```diff
@@ -151,16 +157,6 @@ def call(
...
-
-    Args:
-        feature_extractor: the backbone serving as feature extractor
-        vocab_size: size of the alphabet
-        rnn_units: number of hidden units in both encoder and decoder LSTM
-        embedding_units: number of embedding units
-        attention_units: number of hidden units in attention module
-        max_length: maximum word length handled by the model
-        num_decoders: number of LSTM to stack in decoder layer
-
```
Comment (fg-mindee) [inline doctr/models/recognition/sar.py:163]: We should keep this for the documentation

File: doctr/models/recognition/sar.py [hunk 3]
```diff
@@ -196,13 +192,21 @@ def __init__(
...
         pooled_features = tf.reduce_max(features, axis=1)  # vertical max pooling
         encoded = self.encoder(pooled_features, **kwargs)
-        decoded = self.decoder(features, encoded, **kwargs)
+        if training:
+            if labels:
+                decoded = self.decoder(features, encoded, labels, **kwargs)
+            else:
+                raise ValueError('Need to provide labels during training for teacher forcing')
+        else:
+            decoded = self.decoder(features, encoded, **kwargs)
```
Comment (fg-mindee) [inline doctr/models/recognition/sar.py:209]: you need to pass down `training=training` to the decoder call if you do this

File: test/test_models_recognition.py [hunk 4]
```diff
@@ -52,6 +52,23 @@ def test_recognition_models(arch_name, input_shape, output_size):
...
+    values = tf.random.uniform(shape=[11], minval=0, maxval=118, dtype=tf.dtypes.int64)
+    input_labels = tf.sparse.reorder(
+        tf.sparse.SparseTensor(indices=indices, values=values, dense_shape=[batch_size, 41])
+    )
+    out = reco_model(input_tensor, labels=input_labels, training=True)
+    assert isinstance(out, tf.Tensor)
+    assert isinstance(reco_model, tf.keras.Model)
+    assert out.numpy().shape == (batch_size, *output_size)
+
+
```
Comment (fg-mindee) [inline test/test_models_recognition.py:71]: We'll have to refactor this to test the training call of all recognition models as well. But we can keep this for another PR!

Mind adding a few comments about the input_labels generation?

**Review by fg-mindee — COMMENTED** [review]
Nice PR! I added a few comments

File: doctr/models/recognition/sar.py (hunk 3 above)
Comment (charlesmindee) [reply doctr/models/recognition/sar.py:209]: why ? I think it is useless because I already checked inside the sar if we are in training mode

**Review by fg-mindee — APPROVED** [review]
Thanks for the edits!
